torch/csrc/jit/script/compiler.cpp

```diff
@@ -166,28 +170,24 @@ static bool meaningfulName(const std::string& name) {
 // explicitly scoped language as we descend down nested control structures in
 // the frontend (which themselves don't introduce scopes)
 //
-// The algorithm is roughly as follows:
-// 1) While emitting a block within a control operator, add inputs and outputs
-//      from the block for each value referenced (both "reads" and "writes").
-//      This sets the value up as a candidate loop carried dependency.
-// 2) When we reach the end of the block, examine all the values in the current
-//      scope's value map. If the name also resides in an outer scope with a
-//      different Value*, this is a true loop-carried dependency. If not, this
-//      value was not assigned to. Replace all references to the block input
-//      with the Value* pointed to in the tightest enclosing scope. Then delete
-//      that block input and output.
-// 3) When we emit the actual control operator, take all of the loop-carried
-//      dependency values as inputs and return them as outputs from the control
-//      op
+// The Environment keeps track of two tables, one for values which are first
+// class and another for values which are not. When a first class value
+// is set in the environment, we emit a prim::Store which sets the
+// name of the variable to the type of a value. When a first-class value is
+// referenced we emit a prim::Load that generates a value of the appropriate
+// type type.
 //
-//  Note that an alternative implementation could only add the loop-carried dep
-//      inputs and outputs when we see a value that is mutated. This, however
-//      requires replacing all references to that value *within the current
-//      block* with a new input. That is to say: we need to traverse the pre-
-//      decessor nodes and replace inputs that reference that value with the
-//      newly-created input. This could be made less expensive with a change to
-//      the IR API, but for now we choose to pessimisitically create inputs and
-//      delete unnecessary ones later with replaceAllusesWith().
+// a = 1
+// print(a)
+// becomes:
+// = prim::Store[name="a"](%a.1)
+// %a : int = prim::Load[name="a"]()
+// prim::Print(%a)
+//
+// When we emit a control operator, for all mutated variables we emit a
```

Comment:
I am not a fan of this approach. It is mixing metaphors. Before SSA blocks should not have inputs/outputs for tracking values. It prevents other passes from being able to use Load/Store as well.